### PR TITLE
Added the Google Speech dataset

### PR DESCRIPTION
**Pull request by MRandl**
- Added the Google Speech dataset.
- Fixed various bugs related to chunking.

### CODE REVIEW

File: accdfl/core/model_evaluator.py [hunk 1]
```diff
@@ -58,5 +62,6 @@ def evaluate_accuracy(self, model):
                 num_batches += 1
 
         accuracy = float(correct) / float(example_number) * 100.0
+        print(accuracy, flush = True)
```
Comment (devos50) [inline accdfl/core/model_evaluator.py:65]: This print should probably be removed before merging 😅

**Comment by devos50** [comment]
Hey @MRandl!

I just had a look at this PR, and tested out the code myself. To do so, I reinstantiated a `run.py` script which allows for standalone, local training (e.g., centralized training with only a single peer). Running this on the RCP cluster gives me the following accuracies:

```
dataset,algorithm,peer,peers,round,learning_rate,accuracy,loss
google_speech,standalone,0,1,20,0.050000,6.135654,0.051606
google_speech,standalone,0,1,40,0.050000,15.127175,0.047316
google_speech,standalone,0,1,60,0.050000,18.607764,0.047257
google_speech,standalone,0,1,80,0.050000,20.995091,0.040531
google_speech,standalone,0,1,100,0.050000,13.833110,0.050050
google_speech,standalone,0,1,120,0.050000,35.720660,0.036274
google_speech,standalone,0,1,140,0.050000,36.746988,0.036643
google_speech,standalone,0,1,160,0.050000,40.294511,0.035781
google_speech,standalone,0,1,180,0.050000,32.842481,0.037730
google_speech,standalone,0,1,200,0.050000,19.009371,0.049966
google_speech,standalone,0,1,220,0.050000,36.702365,0.037544
google_speech,standalone,0,1,240,0.050000,46.050870,0.030653
google_speech,standalone,0,1,260,0.050000,33.757251,0.037043
google_speech,standalone,0,1,280,0.050000,37.773315,0.036695
google_speech,standalone,0,1,300,0.050000,39.022758,0.033988
google_speech,standalone,0,1,320,0.050000,48.795181,0.029974
google_speech,standalone,0,1,340,0.050000,26.483713,0.043957
google_speech,standalone,0,1,360,0.050000,54.663097,0.027995
```

So, it seems to be working in a centralized setting: reaching 54.6% accuracy after 360 rounds is quite decent for a classification problem with 35 classes, and accuracy is still improving 😀.

I will test to see if it works with Conflux next. Also, I was wondering if you are adding the background noise to the samples? I believe other implementations did this as well.

**Comment by devos50** [comment]
Update: I just tested it with Conflux and 10 peers (IID data distribution). These are my accuracies:

```
dataset,seed,learning_rate,group,time,peer,round,accuracy,loss
google_speech,42,0.05,"s=10",4.070000,0,10,0.200803,0.061260
google_speech,42,0.05,"s=10",4.070000,2,10,0.200803,0.061260
google_speech,42,0.05,"s=10",4.070000,7,10,0.200803,0.061260
google_speech,42,0.05,"s=10",4.070000,1,10,0.200803,0.061260
google_speech,42,0.05,"s=10",4.070000,6,10,0.200803,0.061260
google_speech,42,0.05,"s=10",4.070000,4,10,0.200803,0.061260
google_speech,42,0.05,"s=10",4.070000,8,10,0.200803,0.061260
google_speech,42,0.05,"s=10",4.070000,3,10,0.200803,0.061260
google_speech,42,0.05,"s=10",4.070000,5,10,0.200803,0.061260
google_speech,42,0.05,"s=10",4.070000,9,10,0.200803,0.061260
google_speech,42,0.05,"s=10",7.770000,1,20,0.200803,0.070738
google_speech,42,0.05,"s=10",7.770000,3,20,0.200803,0.070738
google_speech,42,0.05,"s=10",7.770000,7,20,0.200803,0.070738
google_speech,42,0.05,"s=10",7.770000,9,20,0.200803,0.070738
google_speech,42,0.05,"s=10",7.770000,6,20,0.200803,0.070738
google_speech,42,0.05,"s=10",7.770000,4,20,0.200803,0.070738
google_speech,42,0.05,"s=10",7.770000,2,20,0.200803,0.070738
google_speech,42,0.05,"s=10",7.770000,5,20,0.200803,0.070738
google_speech,42,0.05,"s=10",7.770000,8,20,0.200803,0.070738
google_speech,42,0.05,"s=10",7.770000,0,20,0.200803,0.070738
google_speech,42,0.05,"s=10",11.470000,4,30,8.969210,0.115758
google_speech,42,0.05,"s=10",11.470000,5,30,8.969210,0.115758
google_speech,42,0.05,"s=10",11.470000,9,30,8.969210,0.115758
google_speech,42,0.05,"s=10",11.470000,2,30,8.969210,0.115758
google_speech,42,0.05,"s=10",11.470000,3,30,8.969210,0.115758
google_speech,42,0.05,"s=10",11.470000,0,30,8.969210,0.115758
google_speech,42,0.05,"s=10",11.470000,1,30,8.969210,0.115758
google_speech,42,0.05,"s=10",11.470000,8,30,8.969210,0.115758
google_speech,42,0.05,"s=10",11.470000,6,30,8.969210,0.115758
google_speech,42,0.05,"s=10",11.470000,7,30,8.969210,0.115758
google_speech,42,0.05,"s=10",15.170000,7,40,8.969210,0.175999
google_speech,42,0.05,"s=10",15.170000,8,40,8.969210,0.175999
google_speech,42,0.05,"s=10",15.170000,3,40,8.969210,0.175999
google_speech,42,0.05,"s=10",15.170000,4,40,8.969210,0.175999
google_speech,42,0.05,"s=10",15.170000,0,40,8.969210,0.175999
google_speech,42,0.05,"s=10",15.170000,1,40,8.969210,0.175999
google_speech,42,0.05,"s=10",15.170000,5,40,8.969210,0.175999
google_speech,42,0.05,"s=10",15.170000,9,40,8.969210,0.175999
google_speech,42,0.05,"s=10",15.170000,2,40,8.969210,0.175999
google_speech,42,0.05,"s=10",15.170000,6,40,8.969210,0.175999
google_speech,42,0.05,"s=10",18.870000,0,50,0.200803,0.256305
google_speech,42,0.05,"s=10",18.870000,5,50,0.200803,0.256305
google_speech,42,0.05,"s=10",18.870000,9,50,0.200803,0.256305
...
```

So it's getting to a few percent accuracy (more than a random guess) and then collapses again to an extremely low accuracy. The loss is going up, which is very weird. I will spend some time debugging this to see what could be causing it.

**Comment by MRandl** [comment]
Hey martijn,
indeed, the background noise is not implemented. I was mostly fighting with the random crashes (which I now think are fixed) and the quite low accuracies on Conflux. I initially though the two were related but it seems not to be the case, as I haven't observed an RCP crash anymore, I think it was mostly the pytorch version I was using.

I can have a dab at this again once the EuroMlSys deadline is behind us

**Comment by devos50** [comment]
Great, thanks for the update!

I just tested it when running with a single peer, and I'm also getting low accuracies. This is very weird because Conflux with a single peer _should_ be equivalent to centralized training. There might be something going on in the chunking mechanism, which I will check later.

There was also a bug when running Conflux with less than 10 peers, I just fixed that as well.

```
dataset,seed,learning_rate,group,time,peer,round,accuracy,loss
google_speech,42,0.05,"s=1",0.000000,0,10,8.969210,0.078716
google_speech,42,0.05,"s=1",0.000000,0,20,8.969210,0.088076
google_speech,42,0.05,"s=1",0.000000,0,30,0.200803,0.189617
google_speech,42,0.05,"s=1",0.000000,0,40,9.482374,0.279529
google_speech,42,0.05,"s=1",0.000000,0,50,8.969210,0.602358
google_speech,42,0.05,"s=1",0.000000,0,60,8.969210,0.799750
google_speech,42,0.05,"s=1",0.000000,0,70,8.969210,0.809868
google_speech,42,0.05,"s=1",0.000000,0,80,8.969210,1.218372
google_speech,42,0.05,"s=1",0.000000,0,90,8.969210,0.801289
```

> I can have a dab at this again once the EuroMlSys deadline is behind us

Great, thanks!

**Comment by devos50** [comment]
Update: it's most likely the chunking mechanism. This is what I get with one peer and when using one chunk (e.g., not chunking the model):

```
dataset,seed,learning_rate,group,time,peer,round,accuracy,loss
google_speech,42,0.05,"s=1",0.000000,0,10,8.969210,0.065479
google_speech,42,0.05,"s=1",0.000000,0,20,8.969210,0.077829
google_speech,42,0.05,"s=1",0.000000,0,30,0.200803,0.198705
google_speech,42,0.05,"s=1",0.000000,0,40,0.200803,0.256714
google_speech,42,0.05,"s=1",0.000000,0,50,8.969210,0.442230
```

Will explore this later. Maybe my chunking method is invalid, we could try the one from Virtual Nodes: https://github.com/sacs-epfl/shatter/blob/main/artifact_scripts/gradientInversion/rog/src/compress/compressor.py#L112

**Comment by devos50** [comment]
The problem seems to be batch normalisation in ResNet models (Thanks Rishi for helping out here). We disabled this as well for Virtual Nodes.

Disabling BN gives the following progression (with LR=0.1):

```
dataset,seed,learning_rate,group,time,peer,round,accuracy,loss
google_speech,42,0.1,"s=1",0.000000,0,10,6.537260,0.055595
google_speech,42,0.1,"s=1",0.000000,0,20,0.401606,0.054672
google_speech,42,0.1,"s=1",0.000000,0,30,2.610442,0.054400
google_speech,42,0.1,"s=1",0.000000,0,40,2.521196,0.054104
google_speech,42,0.1,"s=1",0.000000,0,50,4.908523,0.054166
google_speech,42,0.1,"s=1",0.000000,0,60,4.997769,0.054080
google_speech,42,0.1,"s=1",0.000000,0,70,9.036145,0.053627
google_speech,42,0.1,"s=1",0.000000,0,80,7.340473,0.052925
google_speech,42,0.1,"s=1",0.000000,0,90,11.691209,0.053396
google_speech,42,0.1,"s=1",0.000000,0,100,9.303882,0.053701
google_speech,42,0.1,"s=1",0.000000,0,110,5.377064,0.053016
google_speech,42,0.1,"s=1",0.000000,0,120,6.894244,0.053265
google_speech,42,0.1,"s=1",0.000000,0,130,5.778670,0.052661
google_speech,42,0.1,"s=1",0.000000,0,140,9.906292,0.052469
google_speech,42,0.1,"s=1",0.000000,0,150,7.518965,0.052450
google_speech,42,0.1,"s=1",0.000000,0,160,9.549308,0.052457
google_speech,42,0.1,"s=1",0.000000,0,170,10.307898,0.052438
google_speech,42,0.1,"s=1",0.000000,0,180,6.202588,0.052810
google_speech,42,0.1,"s=1",0.000000,0,190,9.683177,0.052497
google_speech,42,0.1,"s=1",0.000000,0,200,3.279786,0.052887
google_speech,42,0.1,"s=1",0.000000,0,210,5.689424,0.051989
google_speech,42,0.1,"s=1",0.000000,0,220,5.890228,0.051606
google_speech,42,0.1,"s=1",0.000000,0,230,9.393128,0.049517
google_speech,42,0.1,"s=1",0.000000,0,240,12.561357,0.050389
```

It's slow but it's learning at least :)

**Comment by devos50** [comment]
Fixed this by chunking on the `state_dict` instead of parameters: https://github.com/devos50/decentralized-learning/commit/556068baecf2c03b407680011e75d9f295b3ad4e. After some more testing, I'll open a PR on top of this PR, and them merge everything.
